### PR TITLE
Background image src eval and handle environment variables

### PR DESCRIPTION
**Pull request by vinothvino42**

### CODE REVIEW

**Review by vusters — DISMISSED** [review]

File: lib/widget/image.dart [hunk 1]
```diff
@@ -92,12 +94,15 @@ class ImageState extends WidgetState<EnsembleImage> {
       return placeholder;
     }
 
+    ScopeManager? scopeManager = DataScopeWidget.getScope(context);
+    final imgSource = scopeManager?.dataContext.eval(source) ?? source;
```
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/widget/image.dart]: > you reverted this later. All good.

Yeah, I noticed that it's already evaluated. So I just reverted it back

File: lib/util/utils.dart [hunk 2]
```diff
@@ -87,7 +88,15 @@ class Utils {
   static BackgroundImage? getBackgroundImage(dynamic value) {
     if (value is Map) {
       if (value['source'] != null) {
-        return BackgroundImage(value['source'].toString(),
+        Map<String, dynamic>? envOverrides =
```
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/util/utils.dart:91]: The image is not working when we set the background image as an environmentVariable.

```
View:
  # Optional - style the screen
  styles:
    useSafeArea: true
    scrollableView: true
    backgroundImage:
      source: ${env.image_background}
```

[the remark this answers is not in the document]
Comment (vusters) [reply lib/util/utils.dart:91]: Please double check this. page.dart line 282 evaluation should already handle all environment variables. The output of that should already be correct. You shouldn't need all this manual step.

[the remark this answers is not in the document]
Comment (kmahmood74) [reply lib/util/utils.dart:91]: @vusters looks like Utils.getxxx methods do not check for expressions. Check Utils.getColor, Utils.getBackgroundImage - none of them check for expressions and evaluate those expressions. They expect the expressions to be already evaluated. If that's incorrect, we will face the same issue with expressions in backgroundColor, backgroundImage, scrollableView and navigationIconPosition.

Unless expressions are evaluated before the Page.build method, we will need to fix it for all styles to allow for evaluating expressions.

[the remark this answers is not in the document]
Comment (vusters) [reply lib/util/utils.dart:91]: I meant line 282 in this change (above) where Vinoth added the eval (). The evaluation should already handle all the environment variables.

**Comment by kmahmood74** [comment]
@vinothvino42 finish it up please, this has been pending for a while. See Vu's comment, it's a minor change.

File: lib/util/utils.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinothvino42) [reply lib/util/utils.dart]: @vusters I updated the PR. Can you check now?

**Comment by vinothvino42** [comment]
> @vinothvino42 finish it up please, this has been pending for a while. See Vu's comment, it's a minor change.

@vusters @kmahmood74 In the screen.dart
1.
```
Widget renderScreen(ScreenDefinition screenDefinition) {
    // build the data context
    DataContext dataContext = DataContext(
        buildContext: context, initialMap: widget.screenPayload?.arguments);

    PageModel pageModel =
        screenDefinition.getModel(widget.screenPayload?.arguments, dataContext);
```
2
```
class ScreenDefinition {
  ScreenDefinition(this._content);
  final YamlMap _content;

  PageModel getModel(
      Map<String, dynamic>? inputParams, DataContext dataContext) {
    YamlMap output = _content;

    /// manipulate the screen definition to account for custom widget
    if (_content.keys.length == 1 && _content['Widget'] != null) {
      output = _getWidgetAsScreen(_content['Widget'], inputParams);
    }

    final evaluatedData = dataContext.eval(output);
    final data = Utils.getYamlMap(evaluatedData);
    return PageModel.fromYaml(data!, dataContext);
    return PageModel.fromYaml(output, dataContext);
  }
}
```

Here, after evaluating the output (see evaluatedData object). I'm getting null for all the body (code block) of executeCode.

**Comment by kmahmood74** [comment]
@vinothvino42 did you resolve the issue with @vusters today? Any progress on this one?

**Comment by vusters** [comment]
@vinothvino42 this is how you evaluate - all is done for you already:
`dynamic evaluatedBackground = _scopeManager.dataContext.eval(widget._pageModel.pageStyles?['backgroundImage']);`

There should be zero things you have to do. This actually recursively go through the backgroundImage and evaluate all under it.
Regarding evaluating others. Let's not do that yet. The proper solution is to do them all automatically (just like a regular widget), vs go through every single property to evaluate them.

**Pull request merged**